### PR TITLE
Important commit, Analysis change

### PR DESCRIPTION
**Pull request by egolol**
Changed Measurement so it  has an id, arranged everything so it works, also tests. Analysis now has 2 abstract methods and some new functions, arranged everything in unit tests so it works.

### CODE REVIEW

File: braph/analysis/Analysis.m [hunk 1]
```diff
@@ -21,29 +21,75 @@
         end
         % function copyElement() %TODO
     end
+    methods (Abstract, Access = protected)
+        calculateMeasurementID(analysis, varargin)
```
Comment (giovannivolpe) [inline braph/analysis/Analysis.m]: let's call this calcualte_measurement_id() for consistency

File: braph/analysis/Analysis.m [hunk 2]
```diff
@@ -21,29 +21,75 @@
...
+                for i = 1:1:analysis.measurementnumber()
+                    if analysis.getMeasurement(i).getID() == measurement.getID()
+                        bool = true;
+                        break
+                    end
+                end
+            elseif isa(measurement, 'double') % the measurement is the subject_index
+                measurement_index = measurement;
+                bool = isKey(analysis.measurement_dict, measurement_index);
+            else   % measurement is the measurementID.
```
Comment (giovannivolpe) [inline braph/analysis/Analysis.m]: let's write measurement_id for consistency

File: braph/analysis/Analysis.m [hunk 3]
```diff
@@ -21,29 +21,75 @@
...
-                    measurement_index = i;
-                    break;
+            if isa(measurement, 'Measurement')  % if object
+                for i = 1:1:analysis.measurementnumber()
+                    if analysis.getMeasurement(i).getID() == measurement.getID()
+                        measurement_index = i;
+                        break;
+                    end
+                end
+            else  % is the id
```
Comment (giovannivolpe) [inline braph/analysis/Analysis.m]: % measurement is the measuremet_id

File: braph/analysis/Analysis.m [hunk 4]
```diff
@@ -21,29 +21,75 @@
...
+            cohort = analysis.cohort;
+        end
+        function measurement = getNewMeasurement(analysis, id, measurement_class, varargin)
+            measurement = Measurement.getMeasurement(id, measurement_class, ...
+                analysis.getCohort().getBrainAtlases(), ....
+                analysis.getCohort().getGroups(), varargin{:});
+        end
+        function measurement = createMeasurement(analysis, measurement_class, atlases, groups, varargin)
+            measurement = Measurement.getMeasurement(analysis.calculateMeasurementID( ...
+                measurement_class, atlases, groups) ...
```
Comment (giovannivolpe) [inline braph/analysis/Analysis.m]: need to pass also the measurement properties (we need to discuss this, might open an issue to implement it later)

File: braph/analysis/Measurement.m [hunk 5]
```diff
@@ -1,12 +1,15 @@
...
         groups  % cell array with groups
         atlases  % cell array with brain atlases
         settings  % settings of the measurement
         data_dict  % dictionary with data for measurements
     end
     methods (Access = protected)
-        function m = Measurement(atlases, groups, varargin)
+        function m = Measurement(id, atlases, groups, varargin)
+            
+            m.id = id;
```
Comment (giovannivolpe) [inline braph/analysis/Measurement.m]: You might check /enforce that id is a string usign our method tostring()

File: braph/analysis/Measurement.m [hunk 6]
```diff
@@ -112,8 +119,8 @@ function setGroups(m, groups)
             % list of measurments data keys
             datalist = eval([Measurement.getClass(m) '.getDataList()']);
         end
-        function sub = getMeasurement(measurementClass, varargin)
-            sub = eval([measurementClass '(varargin{:})']);
+        function sub = getMeasurement(id, measurementClass, varargin)
```
Comment (giovannivolpe) [inline braph/analysis/Measurement.m]: measurement_class (not measurementClass) for consistency

File: braph/analysis/analyses/AnalysisDTI.m [hunk 7]
```diff
@@ -5,12 +5,33 @@
             analysis = analysis@Analysis(cohort, measurements, varargin{:});
         end
     end
+    methods (Access = protected)
+        function id = calculateMeasurementID(analysis, varargin)
+            n = length(varargin);
+            measurementpart = '';
+            grouppart = '';
+            atlaspart = '';
```
Comment (giovannivolpe) [inline braph/analysis/analyses/AnalysisDTI.m]: not needed

File: braph/analysis/analyses/AnalysisDTI.m [hunk 8]
```diff
@@ -5,12 +5,33 @@
...
+            for i = 1:1:n
+                if isa(varargin{i}, 'BrainAtlas')
+                    atlaspart = strcat(atlaspart, varargin{i}.getName());
+                elseif isa(varargin{i}, 'Group')
+                    grouppart = strcat(grouppart, varargin{i}.getName());
+                elseif isa(varargin{i}, 'char')
+                    measurementpart = varargin{i};
+                end
+            end
+            id = strcat(measurementpart, atlaspart, grouppart);
```
Comment (giovannivolpe) [inline braph/analysis/analyses/AnalysisDTI.m]: let's discuss this together

File: braph/analysis/analyses/AnalysisDTI.m [hunk 9]
```diff
@@ -5,12 +5,33 @@
...
+        end
+        function calculate_measurement(analysis)
+            % ?
```
Comment (giovannivolpe) [inline braph/analysis/analyses/AnalysisDTI.m]: let's discuss also this

File: braph/analysis/analyses/AnalysisMRI.m [hunk 10]
```diff
@@ -5,12 +5,33 @@
             analysis = analysis@Analysis(cohort, measurements, varargin{:});
         end
```
Comment (giovannivolpe) [inline braph/analysis/analyses/AnalysisMRI.m]: same as AnalysisDTI

File: braph/analysis/analyses/AnalysisfMRI.m [hunk 11]
```diff
@@ -5,12 +5,33 @@
             analysis = analysis@Analysis(cohort, measurements, varargin{:});
```
Comment (giovannivolpe) [inline braph/analysis/analyses/AnalysisfMRI.m]: same as AnalysisDTI

File: braph/analysis/measurements/ComparisonMRI.m [hunk 12]
```diff
@@ -68,7 +68,7 @@ function update_groups(m, groups)
             datalist('value') = 'DataScalar';  % all globals for now
         end
         function sub = getMeasurement(measurementClass, varargin)
-            sub = eval([measurementClass '(varargin{:})']);
+            sub = eval([measurementClass '(id, varargin{:})']);
```
Comment (giovannivolpe) [inline braph/analysis/measurements/ComparisonMRI.m]: you need to pass id in the method

File: braph/analysis/measurements/MeasurementMRI.m [hunk 13]
```diff
@@ -72,7 +72,7 @@ function update_groups(m, groups)
             datalist('value') = 'DataScalar';  % all globals for now
         end
         function sub = getMeasurement(measurementClass, varargin)
-            sub = eval([measurementClass '(varargin{:})']);
+            sub = eval([measurementClass '(id, varargin{:})']);
         end
```
Comment (giovannivolpe) [inline braph/analysis/measurements/MeasurementMRI.m]: need to pass id

File: braph/analysis/measurements/RandomComparisonMRI.m [hunk 14]
```diff
@@ -73,7 +73,7 @@ function update_groups(m, groups)
             datalist('value') = 'DataScalar';  % all globals for now
         end
         function sub = getMeasurement(measurementClass, varargin)
```
Comment (giovannivolpe) [inline braph/analysis/measurements/RandomComparisonMRI.m]: need to pass id

File: braph/analysis/Analysis.m (hunk 2 above)
Comment (egolol) [reply braph/analysis/Analysis.m]: measuremente_index is different from measurement_id

**Comment by giovannivolpe** [comment]
Add also function
function group_index = getGroupIndex(cohort, group)
to Cohort

**Comment by egolol** [comment]
this is deprecated. Analysis is solved in #207

**Pull request closed without merging**